### ISSUE TITLE
Can't open & focus editor after dataprovider has been refreshed

### ISSUE THREAD

**Issue by OlliTietavainenVaadin** [vaadin/vaadin-grid-flow#486]
`<vaadin.version>12.0.4</vaadin.version>`
Use case: would like to open the Grid's editor and focus the editor field when new data is added to the Grid so the user can start entering data immediately. Trying to do this will cause an `IllegalStateException: the item is not in the backing data provider` on the server side.

Sample to reproduce the issue:
```
        Grid<Bean> grid = new Grid<>();

        List<Bean> items = new ArrayList<>();
        // some test data
        for (int i = 0; i < 5; i++) {
            Bean bean = new Bean(i+"");
            items.add(bean);
        }
        ListDataProvider<Bean> ldp = new ListDataProvider<>(items);
        grid.setDataProvider(ldp);
        Grid.Column<Bean> nameColumn = grid.addColumn(bean -> bean.getName()).setHeader("name");

        Binder<Bean> binder = new Binder<>(Bean.class);
        Editor<Bean> editor = grid.getEditor();

        editor.setBinder(binder);

        TextField field = new TextField();
        binder.forField(field).bind(Bean::getName, Bean::setName);

        nameColumn.setEditorComponent(field);
        editor.addOpenListener(e -> field.focus());
        grid.addItemClickListener(e -> editor.editItem(e.getItem()));


        Button button = new Button("Click to add item. This should also open the editor and focus on the TextField", e -> {
            Bean bean = new Bean("added from button");
            items.add(bean);
            ldp.refreshAll();
            editor.editItem(bean);
            // this should be unnecessary because of the openListener
            field.focus();
        });

        add(new Span("Click Grid item to open editor. Field should be focused immediately (this works)."),
                grid,
                button);
```

**Comment by pleku** [comment]
Was a bit confused with the description and code, but now I think I understood the issue, and yes it is a valid bug. Focusing has nothing to do with this, it will probably break anyway.

The `Grid` doesn't know the item until the call for the data provider has been made, and this is not done until the `beforeClientResponse` phase for all attached components. 

Thus the proper fix for this would be to cache the item that the editor should be opened for if it is not present at the point when the editor is opened, and then during each query check whether any of the new items should have their editor triggered.

**Comment by denis-anisimov** [comment]
I have no idea whether it's fixed or not.

**Comment by Klaudeta** [comment]
This issue will be fixed when both PRs #533 and #531 will be merged.

**Comment by pleku** [comment]
This should be picked to V12 version of the Grid, `2.1` branch
